helix-term/src/commands.rs

```diff
@@ -1798,6 +1807,35 @@ fn append_to_line(cx: &mut Context) {
     doc.set_selection(view.id, selection);
 }
 
+// Creates an LspCallback that waits for formatting changes to be computed. When they're done,
+// it applies them, but only if the doc hasn't changed.
+fn make_format_callback(
```

Comment:
Is there any way to cancel this? Rather than getting the formatting changes then check if the document changes have been computed, we cancel the formatting if the document changed. What do you think?

Reply:
That sounds reasonable, but even if we cancel I think we will need to check if the document changed, right? Otherwise, there could be a race condition.

But anyway, is it ok if I leave this to a future change? It seems like it's worth having a general mechanism for cancelling obsolete lsp callbacks, but that also seems like a bigger project...

Reply:
There's this work_done_token that can also be used for cancellation: https://github.com/helix-editor/helix/blob/2dba228c766e5e9e4b0fe0c12d30909f51faa28e/helix-lsp/src/client.rs#L560

I agree that it doesn't need to happen in this PR.

Reply:
Ok, then at least I'll mention this in a comment

Reply:
Yeah, don't need to be done in this PR. But at least we could add a TODO.